src/Nethermind/Nethermind.Synchronization/FastBlocks/SyncStatusList.cs

```diff
@@ -30,43 +30,43 @@ public void GetInfosForBatch(BlockInfo?[] blockInfos)
             int collected = 0;
 
             long currentNumber = LowestInsertWithoutGaps;
-            lock (_statuses)
+            while (collected < blockInfos.Length && currentNumber != 0)
             {
-                while (collected < blockInfos.Length && currentNumber != 0)
+                if (blockInfos[collected] is not null)
                 {
-                    if (blockInfos[collected] is not null)
-                    {
-                        collected++;
-                        continue;
-                    }
+                    collected++;
+                    continue;
+                }
 
-                    switch (_statuses[currentNumber])
-                    {
-                        case FastBlockStatus.Unknown:
-                            blockInfos[collected] = _blockTree.FindCanonicalBlockInfo(currentNumber);
+                switch (_statuses[currentNumber])
+                {
+                    case FastBlockStatus.Unknown:
+                        blockInfos[collected] = _blockTree.FindCanonicalBlockInfo(currentNumber);
+                        lock (_statuses)
```

Comment:
Shouldn't lock be before the switch? So we read `Unkown` and set `Sent` for sure?
Or is this lock even necessary in the first place? (very tricky question)

Reply:
On the other hand if we didn't pack multiple `FastBlockStatusList` to a byte, maybe we could just Interlock on single array element as a byte. Again bringing the collection size 4x (still manageable though), but reducing locking to almost 0?

Reply:
We could also track not only `LowestInsertWithoutGaps` but also other already in-progress request to try to reduce Interlock contention.